server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -802,69 +822,87 @@ private void runBulkRequestInBatch(
+        // which pipelines to execute for each index request based on previously resolved pipelines.
+        // Inner/outer slots are used to map index requests to their position in the original bulk request list.
         for (DocWriteRequest<?> actionRequest : actionRequests) {
-            IndexRequest indexRequest = TransportBulkAction.getIndexWriteRequest(actionRequest);
-            if (indexRequest == null) {
-                if (counter.decrementAndGet() == 0) {
-                    onCompletion.accept(originalThread, null);
+            List<IndexRequest> childIndexRequests = new ArrayList<>();
+            if (actionRequest instanceof UpdateRequest updateRequest) {
+                childIndexRequests.addAll(updateRequest.getChildIndexRequests());
+            } else {
+                IndexRequest childIndexRequest = TransportBulkAction.getIndexWriteRequest(actionRequest);
+                if (childIndexRequest != null) {
+                    childIndexRequests.add(childIndexRequest);
                 }
-                assert counter.get() >= 0;
-                i++;
-                continue;
             }
 
-            // need to set pipeline of the request as NOOP_PIPELINE_NAME so that when we switch back to the write thread
-            // and invoke doInternalExecute of the TransportBulkAction we will not execute the pipeline again.
-            final String pipelineId = indexRequest.getPipeline();
-            indexRequest.setPipeline(NOOP_PIPELINE_NAME);
-            final String finalPipelineId = indexRequest.getFinalPipeline();
-            indexRequest.setFinalPipeline(NOOP_PIPELINE_NAME);
-            final String systemPipelineId = indexRequest.getSystemIngestPipeline();
-            indexRequest.setSystemIngestPipeline(NOOP_PIPELINE_NAME);
+            if (childIndexRequests.isEmpty()) {
+                slot++;
+                continue;
+            }
 
-            List<IngestPipelineInfo> pipelinesInfoList = new ArrayList<>();
+            // Determine pipelines to execute for all child index requests based on previously resolved pipelines
+            for (int innerSlot = 0; innerSlot < childIndexRequests.size(); innerSlot++) {
+                IndexRequest indexRequest = childIndexRequests.get(innerSlot);
+                // need to set pipeline of the request as NOOP_PIPELINE_NAME so that when we switch back to the write thread
+                // and invoke doInternalExecute of the TransportBulkAction we will not execute the pipeline again.
+                final String pipelineId = indexRequest.getPipeline();
+                indexRequest.setPipeline(NOOP_PIPELINE_NAME);
+                final String finalPipelineId = indexRequest.getFinalPipeline();
+                indexRequest.setFinalPipeline(NOOP_PIPELINE_NAME);
+                final String systemPipelineId = indexRequest.getSystemIngestPipeline();
+                indexRequest.setSystemIngestPipeline(NOOP_PIPELINE_NAME);
```

Comment:
We are setting `NOOP_PIPELINE_NAME` fields before checking if there are pipelines. If a request has no pipeline, we are still mutating it unnecessarily. Instead set the NOOP fields only when pipelinesInfoList is not empty